### PR TITLE
Add support for receiving addrv2

### PR DESCRIPTION
**Pull request by conradoplg**
After Nu5 addrv2 messages can be sent in the network. Therefore, the DNS seeder must be able to accept them.

This adds support for receiving addrv2 messages. It depends on https://github.com/ZcashFoundation/btcd/pull/1 but it's currently pointing to that branch so it should already work.

However, it hasn't been tested yet, thus it's a draft.

Part of ZcashFoundation/zebra#2882

### CODE REVIEW

**Comment by conradoplg** [comment]
I added a test with `btcd` itself, and also tested with Zebra as follows.

- Checked out `coredns`, tag `v1.6.9` (you can also use the docker image with https://github.com/ZcashFoundation/coredns-zcash/pull/11, and skip the next two items)
- Run `go get github.com/zcashfoundation/dnsseeder/dnsseed@addrv2` to get the code in this PR
- Compiled with `make`

- Started a custom Zebra node the patch below applied, listening at port 18234, testnet

<details>

```patch
diff --git a/zebra-network/src/protocol/external/addr/v2.rs b/zebra-network/src/protocol/external/addr/v2.rs
index 009d3b59e..ec9d1244f 100644
--- a/zebra-network/src/protocol/external/addr/v2.rs
+++ b/zebra-network/src/protocol/external/addr/v2.rs
@@ -27,11 +27,8 @@ use crate::{
 #[cfg(any(test, feature = "proptest-impl"))]
 use proptest_derive::Arbitrary;
 
-#[cfg(test)]
 use byteorder::WriteBytesExt;
-#[cfg(test)]
 use std::io::Write;
-#[cfg(test)]
 use zebra_chain::serialization::{zcash_serialize_bytes, ZcashSerialize};
 
 /// The maximum permitted size of the `addr` field in `addrv2` messages.
@@ -114,7 +111,6 @@ pub(in super::super) enum AddrV2 {
 // https://zips.z.cash/zip-0155#deployment
 //
 // And Zebra doesn't use different codecs for different peer versions.
-#[cfg(test)]
 impl From<MetaAddr> for AddrV2 {
     fn from(meta_addr: MetaAddr) -> Self {
         let untrusted_services = meta_addr.services.expect(
@@ -203,7 +199,6 @@ impl AddrV2 {
 // Just serialize in the tests for now.
 //
 // See the detailed note about ZIP-155 activation above.
-#[cfg(test)]
 impl ZcashSerialize for AddrV2 {
     fn zcash_serialize<W: Write>(&self, mut writer: W) -> Result<(), std::io::Error> {
         if let AddrV2::IpAddr {
diff --git a/zebra-network/src/protocol/external/codec.rs b/zebra-network/src/protocol/external/codec.rs
index cea85ec57..33d18698a 100644
--- a/zebra-network/src/protocol/external/codec.rs
+++ b/zebra-network/src/protocol/external/codec.rs
@@ -137,7 +137,7 @@ impl Encoder<Message> for Codec {
             Ping { .. } => b"ping\0\0\0\0\0\0\0\0",
             Pong { .. } => b"pong\0\0\0\0\0\0\0\0",
             Reject { .. } => b"reject\0\0\0\0\0\0",
-            Addr { .. } => b"addr\0\0\0\0\0\0\0\0",
+            Addr { .. } => b"addrv2\0\0\0\0\0\0",
             GetAddr { .. } => b"getaddr\0\0\0\0\0",
             Block { .. } => b"block\0\0\0\0\0\0\0",
             GetBlocks { .. } => b"getblocks\0\0\0",
@@ -250,7 +250,7 @@ impl Codec {
 
                 // Regardless of the way we received the address,
                 // Zebra always sends `addr` messages
-                let v1_addrs: Vec<AddrV1> = addrs.iter().map(|addr| AddrV1::from(*addr)).collect();
+                let v1_addrs: Vec<AddrV2> = addrs.iter().map(|addr| AddrV2::from(*addr)).collect();
                 v1_addrs.zcash_serialize(&mut writer)?
             }
             Message::GetAddr => { /* Empty payload -- no-op */ }
```

  </details>

- Started coredns with `./coredns -conf ../coredns-zcash/coredns/Corefile -dns.port 1053` with the corefile below. If using docker, it will be something like `docker run --rm --network host -v ~/zfnd/coredns-zcash/coredns/:/etc/dnsseeder/ zfnd-seeder:latest -dns.port 1053 -conf /etc/dnsseeder/Corefile`)

```
testnet.myhost {
    dnsseed {
        network testnet
        bootstrap_peers localhost:18234
        crawl_interval 15m
        record_ttl 300
    }
}
```

- `coredns` prints the following

```
[INFO] plugin/dnsseed: Getting addresses from bootstrap peers [localhost:18234]
testnet.myhost.:1053
CoreDNS-1.6.9
linux/amd64, go1.17.3, 17665683-dirty
[INFO] plugin/dnsseed: [2021/11/16 15:05:57] testnet crawl complete, met 12 new peers of 12 in 96 seconds
[INFO] plugin/dnsseed: Starting update timer on testnet. Will crawl every 15.0 minutes.
```

- Got the results with `dig` as follows

```
dig @localhost -p 1053 +noall +answer +multiline testnet.myhost
testnet.myhost.		300 IN A 91.121.88.52
testnet.myhost.		300 IN A 157.245.113.22
testnet.myhost.		300 IN A 116.62.229.19
testnet.myhost.		300 IN A 54.209.49.207
testnet.myhost.		300 IN A 3.112.190.48
testnet.myhost.		300 IN A 209.141.47.197
testnet.myhost.		300 IN A 18.168.221.234
testnet.myhost.		300 IN A 88.99.162.254
testnet.myhost.		300 IN A 135.125.225.50
testnet.myhost.		300 IN A 51.15.89.129
testnet.myhost.		300 IN A 165.22.95.188
```

The returned IPs match IPs reported by Zebra:

image

Wireshark shows `addrv2` messages from Zebra to the DNS seeder:

image

File: go.mod [hunk 1]
```diff
@@ -3,12 +3,14 @@ module github.com/zcashfoundation/dnsseeder
...
 	github.com/caddyserver/caddy v1.0.5
 	github.com/coredns/coredns v1.6.9
 	github.com/miekg/dns v1.1.29
 	github.com/pkg/errors v0.9.1
 )
 
-replace github.com/btcsuite/btcd => github.com/gtank/btcd v0.0.0-20191012142736-b43c61a68604
+// Currently pointing to "addrv2" branch
+// TODO: change to "main-zfnd" branch after https://github.com/ZcashFoundation/btcd/pull/1 is merged
+replace github.com/btcsuite/btcd => github.com/ZcashFoundation/btcd v0.22.0-beta.0.20211116150640-079ebf598ccb
```
Comment (teor2345) [inline go.mod:16]: When we merge https://github.com/ZcashFoundation/btcd/pull/1 , GitHub might delete the branch with this commit, and clean up the commit.

So we either need to:
* update to `main-zfnd` before we merge this PR, or
* tag the commit we're depending on.

File: zcash/client.go [hunk 2]
```diff
@@ -40,7 +40,7 @@ var defaultPeerConfig = &peer.Config{
 	// TODO: fork https://github.com/gtank/btcd/blob/master/peer/peer.go
 	//       and set MinAcceptableProtocolVersion based on the most recently activated network upgrade
 	//       see ticket #10 for details
-	ProtocolVersion:  170015, // Zcash NU5 testnet (second activation)
+	ProtocolVersion: 170015, // Zcash NU5 testnet (second activation)
```
Comment (teor2345) [inline zcash/client.go:43]: ZIP 155 bumps the network protocol version:
https://zips.z.cash/zip-0155#deployment

Unlike Zebra, there's nothing else we need to do to support NU5 in the seeder.

```suggestion
	ProtocolVersion: 170017, // Zcash NU5 mainnet with addrv2
```

**Review by teor2345 — COMMENTED** [review]
This looks good, but we should increment the network protocol version.

- [ ] Can we also document the behaviour of the seeder in a comment?

The intended behaviour is:
* parse received `addr` and `addrv2`
* only send `addr`

To be stricly  compliant with ZIP-155, we should check for a peer version of `170017` before receiving `addrv2`. But that's another edge case we have to implement and test. So I'd prefer to just unconditionally accept `addrv2`, which is compatible behaviour.

(It's also what Zebra does.)

File: go.mod (hunk 1 above)
Comment (conradoplg) [reply go.mod:16]: Agreed, I think it's easier to update to `main-zfnd` just after https://github.com/ZcashFoundation/btcd/pull/1 is accepted.

**Comment by conradoplg** [comment]
> This looks good, ~but we should increment the network protocol version.~
> 
> * [ ]  Can we also document the behaviour of the seeder in a comment?
> 
> The intended behaviour is:
> 
> * parse received `addr` and `addrv2`
> * only send `addr`
> 
> To be stricly compliant with ZIP-155, we should check for a peer version of `170017` before receiving `addrv2`. But that's another edge case we have to implement and test. So I'd prefer to just unconditionally accept `addrv2`, which is compatible behaviour.
> 
> (It's also what Zebra does.)

@teor2345  I added a comment in 7069091606519, but not sure if it's enough. Were you considering some specific place?

**Comment by teor2345** [comment]
> @teor2345 I added a comment in [7069091](https://github.com/ZcashFoundation/dnsseeder/commit/7069091606519dbfcfecf6d6315a17b2cd5255eb), but not sure if it's enough. Were you considering some specific place?

I'd like a code comment that documents how the seeder implements ZIP 155.

I'd also like to document the address sending behaviour.
But I'm not sure what the seeder actually does:
* receives `addr` and `addrv2` messages, and sends `addr` messages
* receives `addr` and `addrv2` messages, and doesn't send any addresses to peers

Can you add code comment where the seeder sends `addr` messages to other peers?
(Or where it disables sending them?)

If that all happens in `btcsuite`, then we can just document it near the protocol version config.

**Review by teor2345 — APPROVED** [review]
Looks great, let's merge!
